TensorFlow autolog automatic run ending

juntai-zheng opened the pull request:
## What changes are proposed in this pull request?

This PR allows autolog to automatically end runs it creates when using `tf.estimator.export_saved_model()`, `tf.estimator.export_savedmodel()`, and `tf.estimator.train()`. In addition, the parameters `steps` and `max_steps` are automatically logged when calling `tf.estimator.train()`.

## How is this patch tested?

Unit tests.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Runs automatically created by `mlflow.tensorflow.autolog()` and `mlflow.keras.autolog()` (#2088) are now automatically ended after training and/or exporting your model. See the docs for more details.

### What component(s) does this PR affect?

- [ ] UI
- [ ] CLI
- [ ] API
- [ ] REST-API
- [ ] Examples
- [ ] Docs
- [X] Tracking
- [ ] Projects
- [ ] Artifacts
- [ ] Models
- [ ] Scoring
- [ ] Serving
- [ ] R
- [ ] Java
- [X] Python

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

juntai-zheng committed:
added keras logic and tests

juntai-zheng committed:
refactored keras autolog tests

juntai-zheng committed:
minor edits to keras tests

juntai-zheng committed:
minor edit to documentation

juntai-zheng committed:
added autolog ending for tf.Keras, added patching for tf.keras.model.fit_generator

juntai-zheng committed:
refactored and added tests for tf.keras

juntai-zheng committed:
merging from upstream

juntai-zheng committed:
removed fit_generator hook for another PR

juntai-zheng committed:
renamed create_model to create_keras_model

juntai-zheng committed:
refactored and added tests for auto ending tf.keras autolog runs

juntai-zheng committed:
made test names more descriptive

juntai-zheng committed:
lint

juntai-zheng committed:
added run starting in keras

juntai-zheng committed:
refactored run starting into yield fixture

juntai-zheng committed:
added logic for run autoending in tf.estimator

juntai-zheng committed:
added logic and tests for tf estimator run ending

juntai-zheng committed:
merging from master

smurching commented on mlflow/tensorflow.py:
```diff
@@ -657,26 +657,67 @@ def autolog(every_n_iter=100):
+    def train(self, *args, **kwargs):
+        global _AUTO_END_RUN
+        if not mlflow.active_run():
+            mlflow.start_run()
```
As per comments on https://github.com/mlflow/mlflow/pull/2095, we should wrap this in a `try_mlflow_log` as well

smurching commented on mlflow/tensorflow.py:
```diff
@@ -657,26 +657,67 @@ def autolog(every_n_iter=100):
+        result = original(self, *args, **kwargs)
+
+        if _AUTO_END_RUN:
+            mlflow.end_run()
```
```suggestion
            try_mlflow_log(mlflow.end_run())
```

smurching commented on mlflow/tensorflow.py:
```diff
@@ -657,26 +657,67 @@ def autolog(every_n_iter=100):
         return
 
+    @gorilla.patch(tensorflow.estimator.Estimator)
+    def train(self, *args, **kwargs):
```
Let's also update the documentation of `autolog` to note that both `train` and `export_saved_model` will start/end runs, and that to have everything autologged under a single run the user should wrap their training logic in a `with mlflow.start_run()`?

smurching replied:
We should also update https://github.com/mlflow/mlflow/blob/master/examples/tensorflow/tf1/train_predict.py and https://github.com/mlflow/mlflow/blob/master/examples/tensorflow/tf2/train_predict_2.py to follow this practice

juntai-zheng replied:
Sure, I'll do the doc updates in this PR and update the examples in a follow-up.

smurching commented on mlflow/tensorflow.py:
```diff
@@ -657,26 +657,67 @@ def autolog(every_n_iter=100):
                        tf_signature_def_key='predict',
                        artifact_path='model')
+        if _AUTO_END_RUN:
+            mlflow.end_run()
```
```suggestion
            try_mlflow_log(mlflow.end_run())
```

smurching commented on mlflow/tensorflow.py:
```diff
@@ -657,26 +657,67 @@ def autolog(every_n_iter=100):
                        tf_signature_def_key='predict',
                        artifact_path='model')
+        if _AUTO_END_RUN:
+            mlflow.end_run()
```
```suggestion
            try_mlflow_log(mlflow.end_run())
```

smurching commented on tests/tensorflow_autolog/test_tensorflow2_autolog.py:
```diff
@@ -182,24 +182,26 @@ def input_fn(features, labels, training=True, batch_size=256):
 @pytest.mark.large
+@pytest.mark.parametrize('export', [True])
 def test_tf_estimator_autolog_model_can_load_from_artifact(tf_estimator_random_data_run):
     artifacts = client.list_artifacts(tf_estimator_random_data_run.info.run_id)
```
Hm how does this test work? As it stands, it only pass if we have `manual_run=True` (because otherwise the run we log metrics etc & the run we log the saved model under will be different?)

juntai-zheng replied:
It works because the fixture `tf_estimator_random_data_run` would return the most recent run, where in both cases of `manual_run=True` and `manual_run=False` would return the run where the model is logged.

smurching reviewed (commented):
BTW, here's another way we could implement the auto-run-ending for TF in a way that could allow us to log to the same run even if the user does `export_saved_model`: we keep track of the last run ID created by autolog, i.e. we have a global `_AUTOLOG_RUN_ID` initialized to `None`, set it to the ID of the run started in `train` if there's no active run, and then resume that run in `export_saved_model` by calling `mlflow.start_run(run_id=_AUTOLOG_RUN_ID)` instead of just `mlflow.start_run()`. Could be worth doing that actually, I don't think the code change is too large, and then we'd continue to have the nice behavior of things "just working" if you add `mlflow.tensorflow.autolog`, without needing to recommend wrapping your code in a `mlflow.start_run`

juntai-zheng committed:
added _autolog_run_id, actually patched train

juntai-zheng committed:
fixed wrapper tries, refactored tests

juntai-zheng committed:
forgot a couple of parentheses

juntai-zheng committed:
updated docs

juntai-zheng committed:
updated docs more, removed extra line in tests

juntai-zheng committed:
changed wording for docs

juntai-zheng committed:
added tests for tf1

juntai-zheng committed:
added space in docs

juntai-zheng committed:
moved space in docs

juntai-zheng committed:
removed code in docs

juntai-zheng committed:
grammar in docs

juntai-zheng committed:
added param checking test

juntai-zheng committed:
Merge branch 'master' of https://github.com/mlflow/mlflow into tf-estimator-autoend

smurching commented on mlflow/tensorflow.py:
```diff
@@ -55,6 +55,7 @@
 
 # For tracking if the run was started by autologging.
 _AUTO_END_RUN = False
```
Could we just remove this global & check whether `_AUTOLOG_RUN_ID` is `None` in our logic below?

juntai-zheng replied:
Yup, now that you mention it I see now I can refactor it.

smurching commented on tests/tensorflow_autolog/test_tensorflow_autolog.py:
```diff
@@ -167,68 +167,101 @@ def test_tf_core_autolog_logs_scalars(tf_core_random_tensors):
-            model_dir=dir)
+@pytest.mark.large
+@pytest.mark.parametrize('export', [True, False])
+def test_tf_keras_autolog_persists_manually_created_run(export):
```
BTW, I'd use pytest's [`tmpdir` fixture](http://doc.pytest.org/en/latest/tmpdir.html) here, which handles automatically creating & deleting a tempdir for you

juntai-zheng replied:
Totally forgot about that feature, will put it in. Thanks for the catch!

juntai-zheng committed:
fixed malformed table

juntai-zheng committed:
removing _AUTO_END_RUN

juntai-zheng committed:
merging from master

juntai-zheng committed:
added check for ending auto runs

juntai-zheng committed:
replaced manual tempdir with tmpdir fixture

smurching commented on docs/source/tracking.rst:
```diff
@@ -217,20 +217,30 @@ log statements. See example usages with `Keras <https://github.com/mlflow/mlflow
 
+If no active run exists when `autolog()` captures data, MLflow will automatically create a run to log information to.
+Once training ends via calls to `tf.estimator.train()`, `tf.keras.fit()`, `tf.keras.fit_generator()`, `keras.fit()` or `keras.fit_generator()`,
+or once `tf.estimator` models are exported via `tf.estimator.export_saved_model()` or `tf.estimator.export_savedmodel()`, MLflow will automatically end that run.
```
```suggestion
or once `tf.estimator` models are exported via `tf.estimator.export_saved_model()`, MLflow will automatically end that run.
```

(`tf.estimator.export_savedmodel` was deprecated in the last TF 1.x release, so I don't think we need to mention it here)

smurching commented on docs/source/tracking.rst:
```diff
@@ -217,20 +217,30 @@ log statements. See example usages with `Keras <https://github.com/mlflow/mlflow
+Once training ends via calls to `tf.estimator.train()`, `tf.keras.fit()`, `tf.keras.fit_generator()`, `keras.fit()` or `keras.fit_generator()`,
+or once `tf.estimator` models are exported via `tf.estimator.export_saved_model()` or `tf.estimator.export_savedmodel()`, MLflow will automatically end that run.
+
+**Note**: When a run is automatically created by `mlflow.tensorflow.autolog`, a call to `tf.estimator.train()` will end the current run but will not log the trained model to that run.
```
Maybe we can reword this more concisely as:

```suggestion
**Note**: When using the `tf.Estimator()` API, `mlflow.tensorflow.autolog` will record training metrics & params passed to `train`. Autologging also logs the models saved by `tf.estimator.export_saved_model()` to MLflow.
```

smurching commented on mlflow/tensorflow.py:
```diff
@@ -584,8 +584,8 @@ def _log_event(event):
     """
     if not mlflow.active_run():
         try_mlflow_log(mlflow.start_run)
```
Should we pull this logic for checking if a run is active & creating one if not into a helper like `_maybe_create_active_run`? Would make it easier to standardize

smurching commented on mlflow/tensorflow.py:
```diff
@@ -657,34 +657,82 @@ def autolog(every_n_iter=100):
+
+        result = original(self, *args, **kwargs)
+
+        if mlflow.active_run().info.run_id == _AUTOLOG_RUN_ID:
```
```suggestion
        if mlflow.active_run() is not None and mlflow.active_run().info.run_id == _AUTOLOG_RUN_ID
```

Just to be safe. Also, we could pull this into a helper as well (e.g. `_maybe_end_active_run`) to reduce duplication

smurching replied:
Actually, you could wrap both the run-creation & run-ending logic in a `contextmanager`, e.g.

```
    @contextmanager
    def _manage_active_run():
        if not mlflow.active_run():
            try_mlflow_log(mlflow.start_run)
        yield mlflow.active_run()
        if mlflow.active_run() is not None and mlflow.active_run().info.run_id == _AUTOLOG_RUN_ID:
            try_mlflow_log(mlflow.end_run)        
```

and then use it via `with _manage_active_run(): ... # do stuff in with block`

https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager has more info on this pattern, we also use it in https://github.com/mlflow/mlflow/blob/690c7e933f4c1007f477fa2b240631a9f8f515cf/mlflow/store/db/utils.py#L68

juntai-zheng replied:
Looks good to me, I just changed the function a little to the following:

```
    @contextmanager
    def _manage_active_run():
        if not mlflow.active_run():
            try_mlflow_log(mlflow.start_run)
            global _AUTOLOG_RUN_ID
            if mlflow.active_run() is not None: # defensive check in case `mlflow.start_run` fails
                _AUTOLOG_RUN_ID = mlflow.active_run().info.run_id
        yield mlflow.active_run()
        if mlflow.active_run() is not None and mlflow.active_run().info.run_id == _AUTOLOG_RUN_ID:
            try_mlflow_log(mlflow.end_run)  
```
This is so `_AUTOLOG_RUN_ID` wouldn't be set unless the run is created by the first `if` statement.

smurching commented on mlflow/tensorflow.py:
```diff
@@ -657,34 +657,82 @@ def autolog(every_n_iter=100):
                        tf_meta_graph_tags=[tag_constants.SERVING],
                        tf_signature_def_key='predict',
                        artifact_path='model')
+        if mlflow.active_run().info.run_id == _AUTOLOG_RUN_ID or auto_end:
```
```suggestion
        if (mlflow.active_run() is not None and mlflow.active_run().info.run_id == _AUTOLOG_RUN_ID) or auto_end:
```

juntai-zheng committed:
removing export_savedmodel from docs

Co-Authored-By: Siddharth Murching <smurching@gmail.com>

juntai-zheng committed:
Update docs wording

Co-Authored-By: Siddharth Murching <smurching@gmail.com>

juntai-zheng committed:
Added defensive check to run ending

Co-Authored-By: Siddharth Murching <smurching@gmail.com>

juntai-zheng committed:
created contextmanager

juntai-zheng committed:
pulling changes
Merge branch 'tf-estimator-autoend' of github.com:juntai-zheng/mlflow into tf-estimator-autoend

juntai-zheng committed:
fixed defensive check

juntai-zheng committed:
lint

smurching reviewed (approved):
LGTM thanks for the updates @juntai-zheng!

smurching commented on docs/source/tracking.rst:
```diff
@@ -217,20 +217,30 @@ log statements. See example usages with `Keras <https://github.com/mlflow/mlflow
+or once `tf.estimator` models are exported via `tf.estimator.export_saved_model()`, MLflow will automatically end that run.
+
+**Note**: When using the `tf.Estimator()` API, `mlflow.tensorflow.autolog` will record training metrics & params passed to `train`. Autologging also logs the models saved by `tf.estimator.export_saved_model()` to MLflow.
+To track a model in the same run, you will need to call `tf.estimator.export_saved_model()`, which will start the most recent `autolog`-created run
```
```suggestion
```

smurching commented on docs/source/tracking.rst:
```diff
@@ -217,20 +217,30 @@ log statements. See example usages with `Keras <https://github.com/mlflow/mlflow
+
+**Note**: When using the `tf.Estimator()` API, `mlflow.tensorflow.autolog` will record training metrics & params passed to `train`. Autologging also logs the models saved by `tf.estimator.export_saved_model()` to MLflow.
+To track a model in the same run, you will need to call `tf.estimator.export_saved_model()`, which will start the most recent `autolog`-created run
+and log the model there. The run will be automatically ended afterwards.
```
```suggestion
```

juntai-zheng committed:
edited docs
